ImportPathMismatchError when on UNC path

Windows: When running pytest on a directory mounted as UNC path, this fail with ImportPathMismatchError:
```
c:\vagrant>pytest tests\\functional\\test_path_encodings.py
pytest tests\\functional\\test_path_encodings.py
Adding c:\python38\lib\site-packages\pywin32_system32 to sys.path
ImportError while loading conftest 'c:\vagrant\tests\functional\conftest.py'.
_pytest.pathlib.ImportPathMismatchError: ('conftest', 'c:\\vagrant\\tests\\functional\\conftest.py', WindowsPath('c:/vagrant/tests/functional/conftest.py'))
```

# How to reproduce
* Fire up a virtual maschine usine Vagrat and sharing the current directory (which contains the package)
* Intall Python, pytest and requiremnets
* Try running pytest

You can actually use the setup provided by PyInstaller:
```
git clone https://github.com/pyinstaller/pyinstaller.git
cd bootloader
# in Vagrantfile, line 434 (provision step "install packages") remove "vcbuildtools"
# … to save avoid installing this huge package
vagrant up --no-provision windows10
vagrant ssh windows10  # password is "Passw0rd!"
pip install -U pip
python -m pip install --upgrade pip setuptools wheel
pip install -U -r tests/requirements-tools.txt
pip install -e .
pytest tests/functional/test_path_encodings.py
```

C:\Users\IEUser>pip list 
Package                   Version   Location
------------------------- --------- ------------------
altgraph                  0.17
apipkg                    1.5
atomicwrites              1.4.0
attrs                     20.1.0
colorama                  0.4.3
execnet                   1.7.1
flake8                    3.8.3
future                    0.18.2
iniconfig                 1.0.1
lxml                      4.5.2
mccabe                    0.6.1
more-itertools            8.4.0
packaging                 20.4
pefile                    2019.4.18
pip                       20.2.2
pluggy                    0.13.1
psutil                    5.7.2
py                        1.9.0
pycodestyle               2.6.0
pyflakes                  2.2.0
pyinstaller               4.1.dev0  \\vboxsvr\vagrant\
pyinstaller-hooks-contrib 2020.7
pyparsing                 2.4.7
pytest                    6.0.1
pytest-drop-dup-tests     0.3.0
pytest-forked             1.3.0
pytest-timeout            1.4.2
pytest-xdist              2.0.0
pywin32                   228
pywin32-ctypes            0.2.0
setuptools                47.1.0
six                       1.15.0
tinyaes                   1.0.1
toml                      0.10.1

System: Windows10, Python 3.8.5

I ran into a new error by the way:

Traceback (most recent call last):
  File "/usr/local/bin/pytest", line 8, in <module>
    sys.exit(console_main())
  File "/usr/local/lib/python3.9/site-packages/_pytest/config/__init__.py", line 185, in console_main
    code = main()
  File "/usr/local/lib/python3.9/site-packages/_pytest/config/__init__.py", line 141, in main
    config = _prepareconfig(args, plugins)
  File "/usr/local/lib/python3.9/site-packages/_pytest/config/__init__.py", line 316, in _prepareconfig
    config = pluginmanager.hook.pytest_cmdline_parse(
  File "/usr/local/lib/python3.9/site-packages/pluggy/hooks.py", line 286, in __call__
    return self._hookexec(self, self.get_hookimpls(), kwargs)
  File "/usr/local/lib/python3.9/site-packages/pluggy/manager.py", line 93, in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
  File "/usr/local/lib/python3.9/site-packages/pluggy/manager.py", line 84, in <lambda>
    self._inner_hookexec = lambda hook, methods, kwargs: hook.multicall(
  File "/usr/local/lib/python3.9/site-packages/pluggy/callers.py", line 203, in _multicall
    gen.send(outcome)
  File "/usr/local/lib/python3.9/site-packages/_pytest/helpconfig.py", line 100, in pytest_cmdline_parse
    config = outcome.get_result()  # type: Config
  File "/usr/local/lib/python3.9/site-packages/pluggy/callers.py", line 80, in get_result
    raise ex[1].with_traceback(ex[2])
  File "/usr/local/lib/python3.9/site-packages/pluggy/callers.py", line 187, in _multicall
    res = hook_impl.function(*args)
  File "/usr/local/lib/python3.9/site-packages/_pytest/config/__init__.py", line 998, in pytest_cmdline_parse
    self.parse(args)
  File "/usr/local/lib/python3.9/site-packages/_pytest/config/__init__.py", line 1272, in parse
    self._preparse(args, addopts=addopts)
  File "/usr/local/lib/python3.9/site-packages/_pytest/config/__init__.py", line 1177, in _preparse
    self.hook.pytest_load_initial_conftests(
  File "/usr/local/lib/python3.9/site-packages/pluggy/hooks.py", line 286, in __call__
    return self._hookexec(self, self.get_hookimpls(), kwargs)
  File "/usr/local/lib/python3.9/site-packages/pluggy/manager.py", line 93, in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
  File "/usr/local/lib/python3.9/site-packages/pluggy/manager.py", line 84, in <lambda>
    self._inner_hookexec = lambda hook, methods, kwargs: hook.multicall(
  File "/usr/local/lib/python3.9/site-packages/pluggy/callers.py", line 208, in _multicall
    return outcome.get_result()
  File "/usr/local/lib/python3.9/site-packages/pluggy/callers.py", line 80, in get_result
    raise ex[1].with_traceback(ex[2])
  File "/usr/local/lib/python3.9/site-packages/pluggy/callers.py", line 187, in _multicall
    res = hook_impl.function(*args)
  File "/usr/local/lib/python3.9/site-packages/pytest_cov/plugin.py", line 125, in pytest_load_initial_conftests
    if early_config.known_args_namespace.cov_source:
AttributeError: 'Namespace' object has no attribute 'cov_source'
Makefile:3: recipe for target 'test' failed
make: *** [test] Error 1

--- a/changelog/7678.bugfix.rst
+++ b/changelog/7678.bugfix.rst
@@ -1,0 +1,2 @@
+Fixed bug where ``ImportPathMismatchError`` would be raised for files compiled in
+the host and loaded later from an UNC mounted path (Windows).

--- a/src/_pytest/pathlib.py
+++ b/src/_pytest/pathlib.py
@@ -543,7 +543,7 @@ def import_path(
             module_file = module_file[: -(len(os.path.sep + "__init__.py"))]
 
         try:
-            is_same = os.path.samefile(str(path), module_file)
+            is_same = _is_same(str(path), module_file)
         except FileNotFoundError:
             is_same = False
 
@@ -551,6 +551,20 @@ def import_path(
             raise ImportPathMismatchError(module_name, module_file, path)
 
     return mod
+
+
+# Implement a special _is_same function on Windows which returns True if the two filenames
+# compare equal, to circumvent os.path.samefile returning False for mounts in UNC (#7678).
+if sys.platform.startswith("win"):
+
+    def _is_same(f1: str, f2: str) -> bool:
+        return Path(f1) == Path(f2) or os.path.samefile(f1, f2)
+
+
+else:
+
+    def _is_same(f1: str, f2: str) -> bool:
+        return os.path.samefile(f1, f2)
 
 
 def resolve_package_path(path: Path) -> Optional[Path]:

--- a/testing/test_pathlib.py
+++ b/testing/test_pathlib.py
@@ -7,6 +7,7 @@ from textwrap import dedent
 import py
 
 import pytest
+from _pytest.monkeypatch import MonkeyPatch
 from _pytest.pathlib import bestrelpath
 from _pytest.pathlib import commonpath
 from _pytest.pathlib import ensure_deletable
@@ -414,3 +415,23 @@ def test_visit_ignores_errors(tmpdir) -> None:
         "bar",
         "foo",
     ]
+
+
+@pytest.mark.skipif(not sys.platform.startswith("win"), reason="Windows only")
+def test_samefile_false_negatives(tmp_path: Path, monkeypatch: MonkeyPatch) -> None:
+    """
+    import_file() should not raise ImportPathMismatchError if the paths are exactly
+    equal on Windows. It seems directories mounted as UNC paths make os.path.samefile
+    return False, even when they are clearly equal.
+    """
+    module_path = tmp_path.joinpath("my_module.py")
+    module_path.write_text("def foo(): return 42")
+    monkeypatch.syspath_prepend(tmp_path)
+
+    with monkeypatch.context() as mp:
+        # Forcibly make os.path.samefile() return False here to ensure we are comparing
+        # the paths too. Using a context to narrow the patch as much as possible given
+        # this is an important system function.
+        mp.setattr(os.path, "samefile", lambda x, y: False)
+        module = import_path(module_path)
+    assert getattr(module, "foo")() == 42

